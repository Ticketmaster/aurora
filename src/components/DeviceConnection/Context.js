import React from "react";

export const { Provider, Consumer } = React.createContext({ "4g": true });
