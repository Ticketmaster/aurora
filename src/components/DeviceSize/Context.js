import React from "react";

export const { Provider, Consumer } = React.createContext({ isSmall: true });

export const withDeviceSize = Component => props => (
  <Consumer>{value => <Component deviceSize={value} {...props} />}</Consumer>
);
