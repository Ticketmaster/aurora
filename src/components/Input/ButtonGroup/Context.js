import React from "react";

export const { Provider, Consumer } = React.createContext({
  selected: null,
  onChange: null
});

export const withButtonGroupConsumer = Component => props => (
  <Consumer>
    {({ onChange: onGroupChange, value: groupValue = {} }) => (
      <Component
        onGroupChange={onGroupChange}
        groupValue={groupValue}
        {...props}
      />
    )}
  </Consumer>
);
