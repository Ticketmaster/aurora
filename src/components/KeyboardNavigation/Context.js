import React from "react";

export const { Provider, Consumer } = React.createContext({
  focused: null,
  focusSelected: () => null
});
