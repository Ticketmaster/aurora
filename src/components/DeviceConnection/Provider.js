import React, { Component } from "react";
import PropTypes from "prop-types";

import { Provider } from "./Context";

import { deviceConnectionStateShape } from "./shape";
import { FALSY_CONNECTIONS_STATE } from "./constants";

class DeviceConnectionProvider extends Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    defaultState: PropTypes.shape(deviceConnectionStateShape)
  };

  static defaultProps = {
    defaultState: false
  };

  constructor(props) {
    super(props);
    const { defaultState } = props;

    this.state = defaultState || this.initialState;
  }

  componentDidMount() {
    const { effectiveType, type } = global.navigator.connection;

    if (effectiveType || type) {
      this.subscribe();
      this.update({ [effectiveType || type]: true, isInitialized: true });
    }
  }

  componentWillUnmount = () => {
    const { isInitialized } = this.state;

    if (isInitialized) {
      this.unsubscribe();
    }
  };

  onConnectionChange() {
    const { effectiveType, type } = global.navigator.connection;
    this.update({ [effectiveType || type]: true });
  }

  initialState = {
    ...FALSY_CONNECTIONS_STATE,
    "4g": true,
    isInitialized: false
  };

  subscribe() {
    global.navigator.connection.addEventListener(
      "change",
      this.onConnectionChange
    );
  }

  // eslint-disable-next-line
  unsubscribe() {
    global.navigator.connection.removeEventListener("change");
  }

  update(payload) {
    this.setState({
      ...FALSY_CONNECTIONS_STATE,
      ...payload
    });
  }

  render() {
    const { children } = this.props;
    return <Provider value={this.state}>{children}</Provider>;
  }
}

export default DeviceConnectionProvider;
