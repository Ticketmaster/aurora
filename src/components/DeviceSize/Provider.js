/* global window */
import React from "react";
import PropTypes from "prop-types";

import { Provider } from "./Context";
import constants from "../../theme/constants";

export default class DeviceSizeProvider extends React.Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    fallbackDetection: PropTypes.func,
    cssOnly: PropTypes.bool
  };

  static defaultProps = { fallbackDetection: null, cssOnly: false };

  initialState = { isSmall: true }; // eslint-disable-line

  state = this.initialState;

  componentDidMount() {
    if (
      typeof window === "undefined" ||
      typeof window.matchMedia !== "function"
    )
      return;

    if (this.props.cssOnly) {
      this.setSize();
      return;
    }

    this.smallMedia = window.matchMedia(constants.breakpoints.small);
    this.mediumMedia = window.matchMedia(constants.breakpoints.medium);
    this.largeMedia = window.matchMedia(constants.breakpoints.large);
    this.xLargeMedia = window.matchMedia(constants.breakpoints.xLarge);
    this.smallMedia.addListener(this.setSize);
    this.mediumMedia.addListener(this.setSize);
    this.largeMedia.addListener(this.setSize);
    this.xLargeMedia.addListener(this.setSize);
    this.setSize();
  }

  componentDidUpdate() {
    if (!this.props.cssOnly) return;

    this.unsubscribe();
  }

  componentWillUnmount() {
    this.unsubscribe();
  }

  setSize = () => {
    if (this.props.cssOnly) {
      this.setState(() => ({ cssOnly: true }));
      return;
    }

    this.setState(() => ({
      isSmall: this.smallMedia.matches,
      isMedium: this.mediumMedia.matches && !this.largeMedia.matches,
      isLarge: this.largeMedia.matches && !this.xLargeMedia.matches,
      isXLarge: this.xLargeMedia.matches,
      cssOnly: false
    }));
  };

  unsubscribe = () => {
    if (this.smallMedia) this.smallMedia.removeListener(this.setSize);
    if (this.mediumMedia) this.mediumMedia.removeListener(this.setSize);
    if (this.largeMedia) this.largeMedia.removeListener(this.setSize);
    if (this.xLargeMedia) this.xLargeMedia.removeListener(this.setSize);
  };

  render() {
    const { fallbackDetection } = this.props;
    const val = fallbackDetection ? fallbackDetection() : this.state;
    return (
      <Provider value={val || this.initialState}>
        {this.props.children}
      </Provider>
    );
  }
}
