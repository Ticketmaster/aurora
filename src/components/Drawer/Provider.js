/* global window: true */
/* eslint react/sort-comp: off, react/no-unused-state: off */
import React, { Children } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

import { Provider } from "./Context";
import { ContentWrapper } from "./Drawer.styles";
import Portal from "../Portal";

export default class DrawerProvider extends React.Component {
  static propTypes = {
    children: PropTypes.element.isRequired,
    className: PropTypes.string
  };

  static defaultProps = { className: "" };

  contentContainer = React.createRef();

  scrollPosition = 0;

  shouldComponentUpdate(prevProps, prevState) {
    return (
      prevState.isOpen !== this.state.isOpen ||
      prevState.content !== this.state.content ||
      prevProps.children !== this.props.children
    );
  }

  componentDidUpdate() {
    const { isOpen } = this.state;
    /* istanbul ignore next */
    if (
      !isOpen &&
      typeof window !== "undefined" &&
      typeof window.scrollTo === "function"
    ) {
      window.scrollTo(0, this.scrollPosition);
    }
  }

  setContent = content => this.setState(() => ({ content }));

  toggleDrawer = () => {
    this.setState(({ isOpen }) => {
      /* istanbul ignore else */
      if (typeof window !== "undefined" && !isOpen) {
        this.scrollPosition = window.pageYOffset;
      }
      return { isOpen: !isOpen };
    });
  };

  state = {
    isOpen: false,
    toggleDrawer: this.toggleDrawer,
    setContent: this.setContent,
    content: null
  };

  render() {
    const { className, children } = this.props;
    const { isOpen, content } = this.state;
    return (
      <Provider value={this.state}>
        <ContentWrapper
          className={classNames({ "drawer--open": isOpen }, className)}
          innerRef={this.contentContainer}
        >
          {Children.only(children)}
          {content && <Portal>{content}</Portal>}
        </ContentWrapper>
      </Provider>
    );
  }
}
