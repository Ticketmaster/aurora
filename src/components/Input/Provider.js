import React from "react";
import PropTypes from "prop-types";
import { Provider } from "./Context";

export default class ItemSelectionProvider extends React.Component {
  static propTypes = {
    children: PropTypes.node.isRequired,
    onChange: PropTypes.func,
    isMultiSelect: PropTypes.bool.isRequired,
    value: PropTypes.arrayOf(PropTypes.string) // eslint-disable-line
  };

  static defaultProps = {
    onChange: null,
    value: []
  };

  static getDerivedStateFromProps(props, state) {
    if (!props.value || !props.value.length) return null;

    if (props.isMultiSelect) {
      if (state.firstUpdate === true) {
        return {
          selected: [...props.value, ...state.selected],
          firstUpdate: false
        };
      }
    }

    if (state.selected.length === 0 && state.firstUpdate === true)
      return { selected: [...props.value] };

    return null;
  }

  componentDidMount() {}

  componentDidUpdate() {
    this.props.onChange && this.props.onChange(this.state.selected); // eslint-disable-line
  }

  onClick = ({ value }) => {
    if (!this.props.isMultiSelect) {
      this.setState({ selected: [value] });
      return;
    }

    if (!this.state.selected.includes(value)) {
      this.setState({ selected: [...this.state.selected, value] });
      return;
    }

    const { selected } = this.state;
    const index = selected.indexOf(value);
    this.setState({
      selected: [...selected.slice(0, index), ...selected.slice(index + 1)]
    });
  };

  state = { selected: [], onClick: this.onClick, firstUpdate: true }; // eslint-disable-line

  render() {
    return <Provider value={this.state}>{this.props.children}</Provider>;
  }
}
