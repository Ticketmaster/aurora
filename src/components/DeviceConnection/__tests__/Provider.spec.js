import React from "react";
import { render, cleanup } from "react-testing-library";

import Provider from "../Provider";
import { Consumer } from "../Context";

describe("DeviceConnectionProvider", () => {
  const addEventListener = jest.fn();
  const removeEventListener = jest.fn();
  const createConnection = jest.fn((data = {}) => ({
    ...data,
    addEventListener,
    removeEventListener
  }));

  beforeEach(() => {
    createConnection.mockClear();
    addEventListener.mockClear();
    removeEventListener.mockClear();
  });

  afterEach(() => {
    cleanup();
  });

  it("should pass the effectiveType to the Consumer's children", () => {
    Object.defineProperty(global.navigator, "connection", {
      writable: true,
      value: createConnection({ effectiveType: "4g" })
    });

    const { container } = render(
      <Provider>
        <Consumer>{val => JSON.stringify(val)}</Consumer>
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it("should listen for a connection change once mounted when the effectiveType API is available", () => {
    Object.defineProperty(global.navigator, "connection", {
      writable: true,
      value: createConnection({ effectiveType: "4g" })
    });

    render(
      <Provider>
        <Consumer>{val => JSON.stringify(val)}</Consumer>
      </Provider>
    );

    expect(addEventListener).toHaveBeenCalled();
  });

  it("should pass the type to the Consumer's children", () => {
    Object.defineProperty(global.navigator, "connection", {
      writable: true,
      value: createConnection({ type: "2g" })
    });

    const { container } = render(
      <Provider>
        <Consumer>{val => JSON.stringify(val)}</Consumer>
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it("should listen for a connection change once mounted when the type API is available", () => {
    Object.defineProperty(global.navigator, "connection", {
      writable: true,
      value: createConnection({ type: "4g" })
    });

    render(
      <Provider>
        <Consumer>{val => JSON.stringify(val)}</Consumer>
      </Provider>
    );

    expect(addEventListener).toHaveBeenCalled();
  });

  it("should not listen for a connection change once mounted when if the browser does not support the API", () => {
    Object.defineProperty(global.navigator, "connection", {
      writable: true,
      value: createConnection()
    });

    render(
      <Provider>
        <Consumer>{val => JSON.stringify(val)}</Consumer>
      </Provider>
    );

    expect(addEventListener).not.toHaveBeenCalled();
  });

  it("should remove listener for a connection change once unmounted when the effectiveType API is available", () => {
    Object.defineProperty(global.navigator, "connection", {
      writable: true,
      value: createConnection({ effectiveType: "4g" })
    });

    const { unmount } = render(
      <Provider>
        <Consumer>{val => JSON.stringify(val)}</Consumer>
      </Provider>
    );

    expect(addEventListener).toHaveBeenCalled();
    unmount();
    expect(removeEventListener).toHaveBeenCalled();
  });

  it("should remove listener for a connection change once unmounted when the type API is available", () => {
    Object.defineProperty(global.navigator, "connection", {
      writable: true,
      value: createConnection({ type: "4g" })
    });

    const { unmount } = render(
      <Provider>
        <Consumer>{val => JSON.stringify(val)}</Consumer>
      </Provider>
    );

    expect(addEventListener).toHaveBeenCalled();
    unmount();
    expect(removeEventListener).toHaveBeenCalled();
  });

  it("should update when the connection changes", () => {
    Object.defineProperty(global.navigator, "connection", {
      writable: true,
      value: createConnection({ effectiveType: "4g" })
    });

    const { container, rerender } = render(
      <Provider>
        <Consumer>{val => JSON.stringify(val)}</Consumer>
      </Provider>
    );

    Object.defineProperty(global.navigator.connection, "effectiveType", {
      writable: true,
      value: "3g"
    });

    rerender(
      <Provider>
        <Consumer>{val => JSON.stringify(val)}</Consumer>
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
