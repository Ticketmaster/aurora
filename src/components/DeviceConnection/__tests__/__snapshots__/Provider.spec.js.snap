// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviceConnectionProvider should pass the effectiveType to the Consumer's children 1`] = `
<div>
  {"slow-2g":false,"2g":false,"3g":false,"4g":true,"isInitialized":true}
</div>
`;

exports[`DeviceConnectionProvider should pass the type to the Consumer's children 1`] = `
<div>
  {"slow-2g":false,"2g":true,"3g":false,"4g":false,"isInitialized":true}
</div>
`;

exports[`DeviceConnectionProvider should update when the connection changes 1`] = `
<div>
  {"slow-2g":false,"2g":false,"3g":false,"4g":true,"isInitialized":true}
</div>
`;
